src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs

```diff
@@ -1224,6 +1226,124 @@ public void Reload_RerunsNamedEndpointConfigurationOnChange()
+        }).Build();
+
+        var mockConfig = new Mock<IConfiguration>();
+        mockConfig.Setup(c => c.GetSection(It.IsAny<string>())).Returns<string>(currentConfig.GetSection);
+        mockConfig.Setup(c => c.GetChildren()).Returns(currentConfig.GetChildren);
+
+        var serverOptions = CreateServerOptions();
+        serverOptions.Configure(mockConfig.Object, reloadOnChange: false);
+        var oldConfigurationLoader = serverOptions.ConfigurationLoader;
+        
+        serverOptions.ConfigurationLoader.Load();
+
+        mockConfig.Verify(c => c.GetSection(It.IsAny<string>()), Times.AtLeastOnce);
+
+        mockConfig.Invocations.Clear();
+
+        serverOptions.Configure(mockConfig.Object, reloadOnChange: false);
+        var newConfigurationLoader = serverOptions.ConfigurationLoader;
+        Assert.NotSame(oldConfigurationLoader, newConfigurationLoader);
+
+        serverOptions.ConfigurationLoader.Load();
+
+        // The change token is stored on the loader, so replacing the loader replaces the token.
+        mockConfig.Verify(c => c.GetSection(It.IsAny<string>()), Times.AtLeastOnce);
+    }
+
+    [Theory]
+    [InlineData(true, true)]
+    [InlineData(false, true)]
+    [InlineData(true, false)]
+    [InlineData(false, false)]
+    public void MultipleLoads_ConfigurationChanges(bool changeBeforeInitialLoad, bool changeAfterInitialLoad)
+    {
+        var currentConfig = new ConfigurationBuilder().AddInMemoryCollection(new[]
+        {
+            new KeyValuePair<string, string>("Endpoints:A:Url", "http://*:5000"),
+            new KeyValuePair<string, string>("Endpoints:B:Url", "http://*:5001"),
+        }).Build();
+
+        var mockReloadToken = new Mock<IChangeToken>();
+        mockReloadToken.SetupGet(t => t.HasChanged).Returns(changeBeforeInitialLoad);
+
+        var mockConfig = new Mock<IConfiguration>();
+        mockConfig.Setup(c => c.GetSection(It.IsAny<string>())).Returns<string>(currentConfig.GetSection);
+        mockConfig.Setup(c => c.GetChildren()).Returns(currentConfig.GetChildren);
+        mockConfig.Setup(c => c.GetReloadToken()).Returns(mockReloadToken.Object);
+
+        var serverOptions = CreateServerOptions();
+        serverOptions.Configure(mockConfig.Object, reloadOnChange: false);
+
+        serverOptions.ConfigurationLoader.Load();
+
+        mockReloadToken.VerifyGet(t => t.HasChanged, Times.Never);
+        mockConfig.Verify(c => c.GetSection(It.IsAny<string>()), Times.AtLeastOnce);
+
+        mockReloadToken.SetupGet(t => t.HasChanged).Returns(changeAfterInitialLoad);
+
+        mockReloadToken.Invocations.Clear();
+        mockConfig.Invocations.Clear();
+
+        serverOptions.ConfigurationLoader.Load();
+
+        mockReloadToken.VerifyGet(t => t.HasChanged, Times.AtLeastOnce);
+        mockConfig.Verify(c => c.GetSection(It.IsAny<string>()), changeAfterInitialLoad ? Times.AtLeastOnce : Times.Never);
```

Comment:
Can we also modify the endpoints between loads assert that `serverOptions.ConfigurationBackedListenOptions` is what we expect after each load?

Reply:
I'm having trouble imagining a failure mode where we get the right number of endpoints but the wrong values, but I'm willing to add additional checks if you think it's important.

Reply:
A check that we have the expected number of endpoints is fine. I don't see that in this test though. The number of endpoints also doesn't change after the first load. The easiest way verify we get the right number is probably to assert the `serverOptions.ConfigurationBackedListenOptions.Count`.

The `MultipleLoads_AddEndpoints` test below verifies that the callbacks are called the correct number of times, but even that doesn't necessarily mean the right number of endpoints have been added. It's easy to verify both.

Reply:
I had assumed `KestrelConfigurationLoader.LocalhostEndpoint` had existing test coverage, but I've updated this test.  Note that these endpoints actually go in `CodeBasedListenOptions`, rather than `ConfigurationBackedListenOptions`.